src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,189 @@
+            {
+                // Most likely a task cancellation exception from the timeout expiring, but regardless we use the handler
+                // dump the raw exception and write the event to the console directly
+                _exceptionhandler(ex);
+                Console.WriteLine(evt);
+            }
+        }
+
+        internal void StartProcessingBuffer(bool batched)
+        {
+            // intentional no-op if the task is already running
+            if (_consoleBufferReadLoop == null || _consoleBufferReadLoop.IsCompleted)
+            {
+                _consoleBufferReadLoop = ProcessConsoleBufferAsync(batched);
+            }
+        }
+
+        private async Task ProcessConsoleBufferAsync(bool batched)
+        {
+            try
+            {
+                if (batched)
+                {
+                    await ProcessConsoleBufferBatchedAsync();
+                }
+                else
+                {
+                    await ProcessConsoleBufferNonBatchedAsync();
+                }
+            }
+            catch (Exception ex)
+            {
+                // Not sure what can fail here, but there is nothing else monitoring this task so just log the exception
+                _exceptionhandler(ex);
+            }
+            finally
+            {
+                // if this has failed for any reason, fall everything back to console
+                _writeEvent = Console.WriteLine;
+                _consoleBuffer.Writer.Complete();
+            }
+        }
+
+        private async Task ProcessConsoleBufferNonBatchedAsync()
+        {
+            await foreach (var line in _consoleBuffer.Reader.ReadAllAsync())
+            {
+                Console.WriteLine(line);
+            }
+        }
+
+        private async Task ProcessConsoleBufferBatchedAsync()
+        {
+            var builder = new StringBuilder();
+
+            while (true)
+            {
+                if (await _consoleBuffer.Reader.WaitToReadAsync() == false)
+                {
+                    // The buffer has been completed and does not allow further reads.
+                    // Currently this should not be possible, but safest thing to do is break out of the loop.
+                    break;
+                }
+
+                if (_consoleBuffer.Reader.TryRead(out string line1))
+                {
+                    // Can we synchronously read multiple lines?
+                    // If yes, use the string builder to batch them together into a single write
+                    // If no, just write the single line without using the builder;
+                    if (_consoleBuffer.Reader.TryRead(out string line2))
+                    {
+                        builder.AppendLine(line1);
+                        builder.AppendLine(line2);
+                        int lines = 2;
+
+                        while (_consoleBuffer.Reader.TryRead(out string nextLine) && lines < DefaultBufferSize)
```

Comment:
`lines < DefaultBufferSize` must go first, otherwise you may dequeue (`TryRead` is `true`), but never append it.

Reply:
great catch, will push a fix.